File: python/julia_client/daemon.py
"""Unix socket server that manages persistent Julia sessions."""

import asyncio
import json
import os
import sys
import time
from pathlib import Path

from .session import SessionManager, DEFAULT_TIMEOUT, PKG_PATTERN

SOCKET_DIR = Path.home() / ".local" / "share" / "julia-client"
DEFAULT_SOCKET = SOCKET_DIR / "julia-daemon.sock"
DEFAULT_IDLE_TIMEOUT = 30 * 60  # 30 minutes

_manager: SessionManager | None = None
_last_request: float = 0.0
_stop_event: asyncio.Event | None = None


async def handle_request(data: dict) -> dict:
    global _last_request
    _last_request = time.monotonic()

    action = data.get("action")

    if action == "eval":
        code = data.get("code", "")
        env_path = data.get("env_path")
        timeout = data.get("timeout")
        julia_cmd = data.get("julia_cmd")

        if timeout is None:
            effective_timeout = None if PKG_PATTERN.search(code) else DEFAULT_TIMEOUT
        else:
            t = float(timeout)
            effective_timeout = t if t > 0 else None

        try:
            session = await _manager.get_or_create(env_path, julia_cmd=julia_cmd)
            output = await session.execute(code, timeout=effective_timeout)
            return {"output": output or "(no output)", "error": None}
        except RuntimeError as e:
            key = _manager._key(env_path)
            if key in _manager._sessions and not _manager._sessions[key].is_alive():
                del _manager._sessions[key]
            return {"output": None, "error": str(e)}

    elif action == "restart":
        await _manager.restart(data.get("env_path"))
        return {"output": "Session restarted.", "error": None}

    elif action == "sessions":
        sessions = _manager.list_sessions()
        if not sessions:
            return {"output": "No active Julia sessions.", "error": None}
        lines = []
        for s in sessions:
            status = "alive" if s["alive"] else "dead"
            label = f"{s['env_path']} (temp)" if s["temp"] else s["env_path"]
            julia = f" julia_cmd={s['julia_cmd']}" if "julia_cmd" in s else ""
            log = f" log={s['log_file']}" if "log_file" in s else ""
            lines.append(f"  {label}: {status}{julia}{log}")
        return {"output": "Active Julia sessions:\n" + "\n".join(lines), "error": None}

    elif action == "stop":
        _stop_event.set()
        return {"output": "Daemon stopping.", "error": None}

    elif action == "ping":
        return {"output": "pong", "error": None}

    else:
        return {"output": None, "error": f"Unknown action: {action!r}"}


async def handle_client(reader: asyncio.StreamReader, writer: asyncio.StreamWriter):
    try:
        raw = await reader.readline()
        if not raw:
            return
        request = json.loads(raw)
        response = await handle_request(request)
        writer.write(json.dumps(response).encode() + b"\n")
        await writer.drain()
    except Exception as e:
        try:
            writer.write(json.dumps({"output": None, "error": f"Daemon error: {e}"}).encode() + b"\n")
            await writer.drain()
        except Exception:
            pass
    finally:
        writer.close()
        try:
            await writer.wait_closed()
        except Exception:
            pass


async def serve(socket_path: Path = DEFAULT_SOCKET, idle_timeout: float = DEFAULT_IDLE_TIMEOUT):
    global _manager, _last_request, _stop_event

    _manager = SessionManager()
    _last_request = time.monotonic()
    _stop_event = asyncio.Event()

    socket_path.parent.mkdir(parents=True, exist_ok=True)
    if socket_path.exists():
        socket_path.unlink()

    pid_path = socket_path.parent / "daemon.pid"

    server = await asyncio.start_unix_server(handle_client, path=str(socket_path))
    pid_path.write_text(str(os.getpid()))
    print(f"julia-daemon listening on {socket_path}", file=sys.stderr)

    async def watchdog():
        while True:
            await asyncio.sleep(10)
            if _stop_event.is_set() or time.monotonic() - _last_request > idle_timeout:
                server.close()
                return

    async with server:
        watchdog_task = asyncio.create_task(watchdog())
        try:
            await server.serve_forever()
        finally:
            watchdog_task.cancel()
            await _manager.shutdown()
            for p in (socket_path, pid_path):
                try:
                    p.unlink()
                except FileNotFoundError:
                    pass


def main():
    import argparse
    parser = argparse.ArgumentParser(prog="julia-daemon", description="Julia REPL daemon")
    parser.add_argument("--socket", type=Path, default=DEFAULT_SOCKET, metavar="PATH")
    parser.add_argument("--idle-timeout", type=float, default=DEFAULT_IDLE_TIMEOUT, metavar="SECS")
    args = parser.parse_args()
    asyncio.run(serve(args.socket, args.idle_timeout))


if __name__ == "__main__":
    main()

File: go/daemon.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"net"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"sync"
	"time"
)

var pkgPattern = regexp.MustCompile(`\bPkg\.`)

type daemonState struct {
	manager     *SessionManager
	mu          sync.Mutex
	lastRequest time.Time
	stopOnce    sync.Once
	stopCh      chan struct{}
}

func handleRequest(state *daemonState, req map[string]any) map[string]any {
	state.mu.Lock()
	state.lastRequest = time.Now()
	state.mu.Unlock()

	action, _ := req["action"].(string)

	switch action {
	case "eval":
		code, _ := req["code"].(string)
		envPath, _ := req["env_path"].(string)
		juliaCmd, _ := req["julia_cmd"].(string)

		var timeoutSecs float64
		if t, ok := req["timeout"]; ok {
			if v, ok := t.(float64); ok && v > 0 {
				timeoutSecs = v
			}
			// v <= 0 means no timeout; timeoutSecs stays 0
		} else if pkgPattern.MatchString(code) {
			timeoutSecs = 0 // Pkg operations: no timeout
		} else {
			timeoutSecs = defaultEvalTimeout
		}

		printResult, _ := req["print_result"].(bool)
		sess, err := state.manager.getOrCreate(envPath, juliaCmd)
		if err != nil {
			return errResp(err.Error())
		}
		output, err := sess.execute(code, timeoutSecs, printResult)
		if err != nil {
			if !sess.isAlive() {
				state.manager.remove(envPath)
			}
			return errResp(err.Error())
		}
		return map[string]any{"output": output, "error": nil}

	case "restart":
		envPath, _ := req["env_path"].(string)
		state.manager.restart(envPath)
		return map[string]any{"output": "Session restarted.", "error": nil}

	case "sessions":
		sessions := state.manager.list()
		if len(sessions) == 0 {
			return map[string]any{"output": "No active Julia sessions.", "error": nil}
		}
		lines := []string{"Active Julia sessions:"}
		for _, s := range sessions {
			status := "alive"
			if !s.alive {
				status = "dead"
			}
			label := s.envPath
			if s.isTemp {
				label += " (temp)"
			}
			line := fmt.Sprintf("  %s: %s", label, status)
			if s.juliaCmd != "" {
				line += " julia_cmd=" + s.juliaCmd
			}
			if s.logFile != "" {
				line += " log=" + s.logFile
			}
			lines = append(lines, line)
		}
		return map[string]any{"output": strings.Join(lines, "\n"), "error": nil}

	case "stop":
		state.stopOnce.Do(func() { close(state.stopCh) })
		return map[string]any{"output": "Daemon stopping.", "error": nil}

	case "ping":
		return map[string]any{"output": "pong", "error": nil}

	default:
		return errResp(fmt.Sprintf("Unknown action: %q", action))
	}
}

func errResp(msg string) map[string]any {
	return map[string]any{"output": nil, "error": msg}
}

func handleConn(conn net.Conn, state *daemonState) {
	defer conn.Close()

	scanner := bufio.NewScanner(conn)
	scanner.Buffer(make([]byte, 1024*1024), 10*1024*1024)
	if !scanner.Scan() {
		return
	}

	var req map[string]any
	if err := json.Unmarshal(scanner.Bytes(), &req); err != nil {
		json.NewEncoder(conn).Encode(errResp(fmt.Sprintf("invalid JSON: %v", err)))
		return
	}

	json.NewEncoder(conn).Encode(handleRequest(state, req))
}

func serveDaemon(socketPath string, idleTimeout time.Duration) error {
	if err := os.MkdirAll(filepath.Dir(socketPath), 0755); err != nil {
		return err
	}
	os.Remove(socketPath)

	ln, err := net.Listen("unix", socketPath)
	if err != nil {
		return err
	}

	pidPath := filepath.Join(filepath.Dir(socketPath), "daemon.pid")
	os.WriteFile(pidPath, []byte(strconv.Itoa(os.Getpid())), 0644)
	fmt.Fprintf(os.Stderr, "julia-daemon listening on %s\n", socketPath)

	state := &daemonState{
		manager:     newSessionManager(),
		lastRequest: time.Now(),
		stopCh:      make(chan struct{}),
	}

	// Idle watchdog: closes listener when idle or stop requested
	go func() {
		ticker := time.NewTicker(10 * time.Second)
		defer ticker.Stop()
		for {
			select {
			case <-state.stopCh:
				ln.Close()
				return
			case <-ticker.C:
				state.mu.Lock()
				idle := time.Since(state.lastRequest)
				state.mu.Unlock()
				if idle > idleTimeout {
					ln.Close()
					return
				}
			}
		}
	}()

	for {
		conn, err := ln.Accept()
		if err != nil {
			break
		}
		go handleConn(conn, state)
	}

	state.manager.shutdown()
	os.Remove(socketPath)
	os.Remove(pidPath)
	return nil
}
